src/psi/j/executors/flux.py

```diff
@@ -0,0 +1,278 @@
+            jpsi_state = JobState.FAILED
+
+        # `exception` events may have error messages (`note`) attached
+        metadata['exit_code'] = exit_code
+        metadata['message'] = metadata.get('note')
+        job_status = JobStatus(jpsi_state, time=time.time(), metadata=metadata)
+        self._update_job_status(jpsi_job, job_status)
+
+    def submit(self, job: Job) -> None:
+
+        with self._lock:
+
+            # derive Flux job spec and submit them
+
+            flux_spec = self._job_2_descr(job)
+
+            # submit the job, register resulting future
+            flux_fut = self._fex.submit(json.dumps(flux_spec))
+            self._jobs[job.id] = [job, flux_fut]
+
+            # asynchroneously wait for the flux jobid
+            jobid_cb = partial(self._jobid_cb, jpsi_id=job.id)
+            flux_fut.add_jobid_callback(jobid_cb)
+
+            # register for event updates
+            for ev in self._event_map:
+                flux_fut.add_event_callback(ev, self._event_cb)
+
+    def _job_2_descr(self, job: Job) -> Dict[str, Any]:
+        jspec = job.spec
+
+        # FIXME: what is spec, what is spec.attributes?
+        # TODO: why don't we get job.attributes default values here?
+        # TODO: why don't we get job.resources default values here?
+        #
+        if jspec.attributes is None:
+            jspec.attributes = JobAttributes()
+
+        if jspec.resources is None:
+            jspec.resources = ResourceSpecV1(process_count=1,
+                                             processes_per_node=1,
+                                             cpu_cores_per_process=1,
+                                             gpu_cores_per_process=1,
+                                             exclusive_node_use=True)
+        if not jspec.executable:
+            raise InvalidJobException('no executable defined')
+
+        if not jspec.arguments:
+            jspec.arguments = []
+
+        rspec = jspec.resources
+        attr = jspec.attributes
+
+        flux_spec = {
+            "tasks": [{
+                "slot": "task",
+                "count": {
+                    "per_slot": 1
+                },
+                "command": [jspec.executable] + jspec.arguments
+            }],
+            "attributes": {
+                "system": {
+                    # FIXME: flux hangs on `None` values
+                    # "cwd": jspec.directory,
+                    "duration": attr.duration.total_seconds()
+                }
+            },
+            "version": 1,
+            "resources": [{
+                "count": rspec.process_count,
+                "type": "slot",
+                "label": "task",
+                "with": [{
+                    "count": rspec.cpu_cores_per_process,
+                    "type": "core"
+                },
+                    # {
+                    #     "count": rspec.gpu_cores_per_process,
+                    #     "type": "gpu"
+                    #  }
```

Comment:
Ah, I wanted to track this down:
```
Job done: JobStatus[FAILED, time=1619781949.7893727, meta={'type': 'alloc', 'severity': 0, 'userid': -1, 'note': "Unsupported resource type 'gpu'", 'exit_code': None, 'message': "Unsupported resource type 'gpu'"}]
```

Do you see anything wrong with the spec above?  But maybe that's just the version of flux I am using here (`flux-core version 0.24.0 - 2021-02-22`)...

Reply:
Ah! So this is a bit weird, but flux-core only ships with `sched-simple`, which is super simple and doesn't support GPUs :(.  So for that jobspec to work, you will need to build and install flux-sched, which includes our graph scheduler `sched-fluxion-resource` that handles GPUs.  If you install flux-core and flux-sched to the same prefix, there should be no extra actions needed on your part. You can confirm that it is loaded with `flux module list`.  Sorry for the complexity there.

Reply:
That makes sense, thanks!

Reply:
Sorry for the late follow-up!  There is still something wrong with the GPU part of the spec.  You were right, installing `flux-sched` along with `flux-core` changed things, but did not resolve the problem  Instead of 
```sh
./tests/run_job.py flux
Job submitted
Job done: JobStatus[FAILED, time=1622363487.4586556, meta={'type': 'alloc', 'severity': 0, 'userid': -1, 'note': "Unsupported resource type 'gpu'", 'exit_code': None, 'message': "Unsupported resource type 'gpu'"}]
```
I now see 
```sh
./tests/run_job.py flux
Job submitted
Job done: JobStatus[FAILED, time=1622363593.7036223, meta={'type': 'alloc', 'severity': 0, 'userid': -1, 'note': 'alloc denied due to type="unsatisfiable"', 'exit_code': None, 'message': 'alloc denied due to type="unsatisfiable"'}]
```
Commenting out the GPU spec part gives me a successful run:
```sh
./tests/run_job.py flux
Job submitted
Job done: JobStatus[COMPLETED, time=1622363659.512951, meta={'status': 0, 'exit_code': 0, 'message': None}]
```

Do you have a suggestion on how to track this down?